### PR TITLE
Feature/coin or cbc interface

### PR DESCRIPTION
**Pull request by braceal**
Here is the PR promised in #212 

Includes: optlang expression parsing bug fix, coinor_cbc interface, coinor_cbc test suite, and unittest interface update (changed assertEquals to assertEqual to remove depreciation warning). 

Special note: I fixed an optlang bug in expression_parsing.py and wrote a short test case exposing it. It has to do with a linear objective with only a constant term. The expression was always being parsed as 0 instead of the constant term. The test case is test_constant_objective in the coinor_cbc test suite (this can be moved to a more appropriate location).

Currently passing 93/103 tests. The failing tests are:

**Name change tests**
ConstraintTestCase::test_change_constraint_name
ModelTestCase::test_change_constraint_name
VariableTestCase::test_change_name
test_changing_variable_names_is_reflected_in_the_solver

_Changing variable/constraint names is not natively supported in MIP (python CBC interface)_

**Dual tests**
test_integer_batch_duals
test_integer_constraint_dual
test_integer_variable_dual

_Just need to know the rule on when to throw an exception_

**Other**
test_indicator_constraint_support
test_timeout - MIP timeout setting is flaky around 0 seconds. Timeout should work for longer problems though

There are a couple small features I would like to add before a merge such as presolve, etc. I am also going to add a few more test cases which solve different example problems. But I wanted to post the code for review.

Documentation also needs to be written.

### CODE REVIEW

**Comment by cdiener** [comment]
This looks good! I wouldn't stress about the mip dependency issue. Python 2.7 and 3.4 are deprecated Python versions anyway and will probably be removed from the CI soon. Do all tests pass now or are you skipping some? The name changing would probably have to work since other tools like cobrapy depend on that. You would usually manage that by an internal mapping of variable/constraint IDs to solver IDs. This way if you change the variable name you would simply update the id in the mapping but maintain the solver ID...

**Comment by cdiener** [comment]
I don't think this should have been closed either...

**Comment by braceal** [comment]
Sorry for my absence. I don't think this should be closed either, the variable/constraint name changing is the last feature to add before it can be released. I've tested it with cobrapy integration and it seems to work fine and this branch is currently being used for a [Flux Balance Analysis app in Kbase](https://narrative.kbase.us/#catalog/apps/COBRApyBasedFBA/run_fba_pipeline/beta). I'm currently on another project, but I can try to have this finished by the end of the year.

If anyone else wants to finish this, please do. I think it is very close.

There are a few other updates that are on a separate branch where I tested the cobrapy, OSQP and COINOR-CBC together. Here is the separate branch: https://github.com/braceal/optlang/tree/test/coinor-cbc_osqp
Here is the compare: https://github.com/braceal/optlang/compare/feature/coin-or-cbc-interface...braceal:test/coinor-cbc_osqp

**Comment by Midnighter** [comment]
Sorry, I know what happened now. I had activated auto-deletion of head branches after merge. @KristianJensen's PR actually merged devel into master. Then devel got deleted and thus all feature branches based on devel got automatically closed and deleted. My apologies. I have removed the setting and will bring back the branches/PRs.

File: src/optlang/coinor_cbc_interface.py
Comment (cdiener) [inline src/optlang/coinor_cbc_interface.py:236]: Make updates a dictionary {var: var * coef / old_coeff} so it works with symengine and sympy.

File: src/optlang/coinor_cbc_interface.py
Comment (cdiener) [inline src/optlang/coinor_cbc_interface.py:709]: Change to `parse_optimization_expression(expr.expand())` to make it work in symengine.

**Review by cdiener — CHANGES_REQUESTED** [review]
Seems to fail due to some incompatibilities with symengine. Might have to adjust tests as well since with symengine you might get expressions like `0.0 + x` which often do not evaluate equal to `x`.

File: src/optlang/coinor_cbc_interface.py
Comment (carrascomj) [reply src/optlang/coinor_cbc_interface.py:709]: `expr.expand()` fails for me since `expr` is an `Objective` or a `Constraint`, which do not have that method. Expanding the internal expression resolved the error at `test_imul_objective`:
```python
    def _expr_to_mip_expr(self, expr):
        """Parses mip linear expression from expression."""
        if hasattr(expr, "expression") and symbolics.USE_SYMENGINE:
            expr._expression = expr.expression.expand()
        offset, coeffs, _ = parse_optimization_expression(expr)
        return offset + mip.xsum(to_float(coef) * self.problem.var_by_name('v_' + var.name)
                                 for var, coef in coeffs.items())
```
Nevertherless, it is still failing because of the dummy comparison at `test_change_objective`:
```
AssertionError: 0.0 + 1.0*ACALD + 1.0*ACALD_reverse_fda2b != 1.0*ACALD + 1.0*ACALD_reverse_fda2b
```

Comment (cdiener) [reply src/optlang/coinor_cbc_interface.py:709]: Just have to call simplify on the expression first.

Comment (carrascomj) [reply src/optlang/coinor_cbc_interface.py:709]: Awesome, calling simplify (on both sides) makes the test suite pass (lines 335 and 340 of `src/optlang/tests/test_coinor_cbc_interface.py`).

Comment (cdiener) [reply src/optlang/coinor_cbc_interface.py:709]: Yeah exactly. This is what we use in cobrapy as well to cover sympy and symengine at the same time.

```Python
def same_ex(ex1, ex2):
    """Compare to expressions for mathematical equality."""
    return ex1.simplify() == ex2.simplify()
```

Comment (carrascomj) [reply src/optlang/coinor_cbc_interface.py:709]: That is handy, definitely. This PR can be merged after applying these two small changes, any thoughts @braceal ?

Comment (braceal) [reply src/optlang/coinor_cbc_interface.py:709]: Hello, Thanks for the activity on this! It would be great to get it merged. 

I remember needing to update the expression parsing logic. Looking back at the code, I believe there was an edge case when a single constant value was used as an expression. Just wanted to make sure this was checked by someone more familiar with the codebase.

https://github.com/opencobra/optlang/pull/213/files#diff-9be3c58a107490c43a136ec465f971e6ef2f76dd5f0080f687cc62e86ae96b05R83

Comment (cdiener) [reply src/optlang/coinor_cbc_interface.py:709]: @braceal If you enable the `Allow edits and access to secrets by maintainers` checkmark on the upper right we could try to fix it directly in this PR.

Comment (braceal) [reply src/optlang/coinor_cbc_interface.py:709]: @cdiener It says it is already enabled. Are you not able to commit changes?
image

Comment (cdiener) [reply src/optlang/coinor_cbc_interface.py:709]: Ups, sorry. Didn't even try. It showed nothing for me so I assumed it's not on but it's just bad UI :sweat_smile:

Comment (KristianJensen) [reply src/optlang/coinor_cbc_interface.py:709]: @braceal Your fix to `expression_parsing` looks fine. Good catch.

Comment (braceal) [reply src/optlang/coinor_cbc_interface.py:709]: Thanks for reviewing @KristianJensen 

What items are remaining before a merge is possible?

Comment (carrascomj) [reply src/optlang/coinor_cbc_interface.py:709]: Let me finish merging #225 and I will make this two couple of changes to pass the CI.

Comment (cdiener) [reply src/optlang/coinor_cbc_interface.py:709]: Probably need to merge the CI fixes from the master branch first (we should rename that one btw).

Comment (carrascomj) [reply src/optlang/coinor_cbc_interface.py:709]: Already merged CI fixes to devel but now OSQP is failing for Windows 3.9 because of https://github.com/oxfordcontrol/osqp/issues/240

Comment (cdiener) [reply src/optlang/coinor_cbc_interface.py:709]: See my comments on the otehr PR and the new one I sent. It's more a version thing. I don't know tox very well though so no idea if it will work.

**Comment by carrascomj** [comment]
MIP [does not support python2](https://github.com/coin-or/python-mip/blob/master/setup.py#L9), so this may be a good chance to consider removing support for it in optlang too.

**Comment by braceal** [comment]
I just recalled, I left a TODO to handle before merge. @cdiener I think this had to do with cobrapy pickle files, do you know what the correct behavior should be here?

https://github.com/braceal/optlang/blob/feature/coin-or-cbc-interface/src/optlang/coinor_cbc_interface.py#L536

**Comment by cdiener** [comment]
That looks good it's the same as in https://github.com/braceal/optlang/blob/feature/coin-or-cbc-interface/src/optlang/osqp_interface.py#L625-L631 right? The configuration needs to pickle the tolerances as well. Cobrapy pickles will be regenerated with new optlang releases so that will work.

**Comment by braceal** [comment]
Ok great, thanks for reviewing! Yes, I agree the logic is the same as in the link you posted. The TODO can probably be removed then.

**Review by cdiener — CHANGES_REQUESTED** [review]
I think this should behave like the other optional solvers. You can use the tox sections to only install mip for Python 3 so it will not run the cbc tests for Python 2.

**Comment by cdiener** [comment]
Fails because there are f-strings in `test_coinor_cbc_interface.py` and that will trigger a syntax error on Python 2 when the file is read.

**Comment by carrascomj** [comment]
Ups, I didn't see that. Working on it!

**Comment by carrascomj** [comment]
If the requested changes were solved for you @cdiener, I think this can be merged.

**Review by cdiener — APPROVED** [review]
Awesome, thanks for carrying this to the finish line.

**Comment by braceal** [comment]
Thank you both so much for seeing this through to completion! @cdiener @carrascomj 🎉

**Pull request merged**
